librosa/segment.py

```diff
@@ -1138,3 +1176,77 @@ def path_enhance(
         np.clip(R_smooth, 0, None, out=R_smooth)
 
     return R_smooth
+
+
+def __affinity_bandwidth(rec, bw_mode, k):
+
+    # the api allows users to specify a scalar bandwidth directly, besides the string based options.
+    if isinstance(bw_mode, (int, float)):
+        bandwidth = bw_mode
+        if bandwidth <= 0:
+            raise ParameterError(
+                "Invalid scalar bandwidth={}. " "Must be strictly positive.".format(bandwidth)
+            )
+        return bandwidth
+
+    if bw_mode is None:
+        bw_mode = 'med_k_scaler'
+
+    if bw_mode not in ['med_k_scaler', 'mean_k', 'gmean_k', 'mean_k_avg', 'gmean_k_avg', 'mean_k_avg_and_pair']:
+        raise ParameterError(
+            (
+                "Invalid bandwidth='{}'. Must be either a positive scaler or one of "
+                "['med_k_scaler', 'mean_k', 'gmean_k', 'mean_k_avg', 'gmean_k_avg', 'mean_k_avg_and_pair']"
+            ).format(bw_mode)
+        )
+
+    # build a list of list that stores the distances to k nearest neighbors for all t points.
+    t = rec.shape[0]
+    knn_dists = []
+    for i in range(t):
+        # Get the links from point i
+        links = rec[i].nonzero()[1]
+
+        # Compute k nearest neighbors' distance and sort ascending
+        knn_dist_row = np.sort(rec[i, links].toarray()[0])[:k]
+
+        # make sure self-links have a distance of 0 and not -1 in self mode
+        knn_dist_row[knn_dist_row < 0] = 0
+        knn_dists.append(knn_dist_row)
+
+    # take the last element of each list for the distance to kth neighbor
+    dist_to_k = np.asarray([dists[-1] for dists in knn_dists])
+    avg_dist_to_first_ks = np.asarray([np.mean(dists) for dists in knn_dists])
```

Comment:
In the case that `sym == True` knocks out some links, should I treat them as 0 distance in my average distance to k? or not include them in the average calculation? The current code compute the means of the links that are left if there are less than `k` links left after taking only mutual k nearest neighbors. The alternative would be `np.sum(dists) / k` in place of `np.mean(dists)`